src/common/authenticator.cpp

```diff
@@ -122,64 +133,230 @@ int Authenticator::HMacSha256(const void* key, int key_size,
     return 0;
 }
 
-std::string Authenticator::Base64(const unsigned char *src, size_t sz) {
-    unsigned char               *pp, *p, *q;
-    unsigned char               *qq = NULL;
-    size_t                      i, safe = sz;
-
-    if (!src || (sz == 0))
-        return (NULL);
-
-    if ((sz % 3) == 1) {
-        p = (unsigned char *)malloc(sz + 2);
-        if (!p)
-            return (NULL);
-        memcpy(p, src, sz);
-        p[sz] = p[sz + 1] = '=';
-        sz += 2;
-    } else if ((sz % 3) == 2) {
-        p = (unsigned char *)malloc(sz + 1);
-        if (!p)
-            return (NULL);
-        memcpy(p, src, sz);
-        p[sz] = '=';
-        sz++;
-    } else {
-        p = (unsigned char *)src;
+std::string Encryptor::Base64Encode(const std::string &src) {
+    return Base64Encode(reinterpret_cast<const unsigned char*>(src.c_str()),
+        src.size());
+}
+
+std::string Encryptor::Base64Encode(const unsigned char *src, size_t sz) {
+    BIO *bio, *b64;
+    BUF_MEM *bufferPtr;
+
+    b64 = BIO_new(BIO_f_base64());
+    BIO_set_flags(b64, BIO_FLAGS_BASE64_NO_NL);
+    bio = BIO_new(BIO_s_mem());
+    bio = BIO_push(b64, bio);
+
+    BIO_write(bio, src, sz);
+    BIO_flush(bio);
+
+    BIO_get_mem_ptr(bio, &bufferPtr);
+    BIO_set_close(bio, BIO_NOCLOSE);
+
+    std::string result;
+    result.assign(bufferPtr->data, bufferPtr->length);
+
+    BIO_free_all(bio);
+    return result;
+}
+
+std::string Encryptor::Base64Decode(const std::string& src) {
+    BIO *bio, *b64;
+    size_t length = src.size();
+    std::unique_ptr<char[]> buffer(new char[length]);
+    std::string decodedString;
+
+    if (length % 4 != 0) {
+        return decodedString;
+    }
+
+    b64 = BIO_new(BIO_f_base64());
+    bio = BIO_new_mem_buf(src.c_str(), length);
+    bio = BIO_push(b64, bio);
+
+    BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
+    int decodedLength = BIO_read(bio, buffer.get(), length);
+
+    if (decodedLength > 0) {
+        decodedString.assign(buffer.get(), decodedLength);
+    }
+
+    BIO_free_all(bio);
+
+    return decodedString;
+}
+
+int Encryptor::AESEncrypt(const std::string &key,
+    const unsigned char *iv, const std::string &in, std::string *out) {
+    // the AES_cbc_encrypt will change iv
+    unsigned char uiv[AES_BLOCK_SIZE];
+    std::memcpy(uiv, iv, AES_BLOCK_SIZE);
+    AES_KEY enc_key;
+    const int keyBits = key.size() * 8;
+    const int pad_len = AES_BLOCK_SIZE - in.size() % AES_BLOCK_SIZE;
+    const int encLength = in.size() + pad_len;
+    // cbc padding
+    std::string data = in;
+    data.resize(data.size() + pad_len, static_cast<char>(pad_len));
+
+    int ret = AES_set_encrypt_key(
+        reinterpret_cast<const unsigned char*>(key.c_str()), keyBits, &enc_key);
+    if (ret != 0) {
+        LOG(ERROR) << "AES_set_encrypt_key failed, error = " << ret;
```

Comment:
It's better not to directly output error logs in common modules

Reply:
So, should we just keep 128? -> All three types can be supported. By default we use 128 length, but users can choose a higher length.

fixed